### DISCUSSION TITLE
D1 Mini 8x8 Matrix LED shield library

### DISCUSSION THREAD

**Discussion by kebwi** [micropython/micropython#9916] [Libraries & Drivers] [answered]
I see a few references to 8x8 led matrices on the old forum, but they appear to target hardware other than the Wemos D1 Mini 8x8 Matrix LED shield, and some of those discussions and sample code utilize very different comms between the microcontroller and the matrix. Some use SPI but seem to indicate pins that differ from the Wemos spec for this particular shield (and I have no reason to expect the messaging protocol to be identical either). Other examples differ so starkly as to use I2C instead of SPI, obviously not very good correspondences to this particular device.

Wemos's webpage for the shield (https://www.wemos.cc/en/latest/d1_mini_shield/matrix_led.html) offers very little information on how to use it. It documents that GPIOs 14 and 13 (D1 mini pins 5 and 7 respectively) are used for CLK and DIN, which conforms to the D1 mini and ESP8266 pinouts of course (merely *implying* that comms are over SPI as opposed to I2C instead of explicitly stating so), but the webpage offers literally no description of the prescribed  messaging protocol once a SPI connection is established. The webpage does, admittedly, link to precisely one Arduino example, and I can slog through translating that to MicroPython and to the distinct pin assignments of the D1 mini and the ESP8266 if I must -- but I thought I would ask here first because I really expected to find that this work has already been done.

One example (https://hackaday.io/project/20839-weather-pal) seems pretty close, but references a library (ht16k33_matrix) that isn't included or linked. That doesn't appear to be quite the same device anyway (this thing https://www.adafruit.com/product/3149 instead of the D1 mini shield) -- and to make matters worse, that appears to be I2C, not SPI, so it seems barely relevant to the D1 mini shield under discussion here.

Despite my failure to find any existing Micropython library or driver for this shield, have I simply overlooked it? I am rather shocked that I would be the first person in the world to obtain this shield and want to use it with a D1 mini (in MicroPython I suppose). Am I truly the first person to ever attempt this? :-) I find that hard to believe.

I'll do it on my own if I have to, and if so, I'll share my code with the group after I create it, but I really expected to find something preexisting for this task, so I just thought I should ask here first.

Thanks!

**Comment by kebwi** [comment, marked as answer]
Of course, after looking all night and then posting the thread above, I found this a few minutes later:

https://forum.micropython.org/viewtopic.php?t=4131

https://github.com/mcauser/micropython-tm1640

I'll see if that works. Seems pretty likely.

**Reply by kebwi** [reply]
Aaaand that worked like a charm. Thanks anyway. Sorry for adding noise to the forum.

**Reply by scruss** [reply]
Better to ask and answer yourself than have no solution.

[Awesome MicroPython](https://awesome-micropython.com/) is a good place to start looking for libraries

**Comment by DaiWenqi** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
